### PR TITLE
feat(skill-sleep): native validation-gated skill self-improvement pipeline (no external dep)

### LINKED ISSUE #32925: Feature Request: Integrate Microsoft SkillOpt for Self-Evolving Agent Skills
### Summary
Microsoft recently released [SkillOpt](https://github.com/microsoft/SkillOpt) — a text-space optimizer that trains reusable natural-language skills for frozen LLM agents through trajectory-driven edits, validation-gated updates, and deployable `best_skill.md` artifacts.

This issue proposes exploring SkillOpt integration with Hermes to enable **automated skill optimization** without touching model weights.

### What is SkillOpt?
- **Paper**: [arXiv:2605.23904](https://arxiv.org/abs/2605.23904)
- **Project Page**: https://microsoft.github.io/SkillOpt/
- **License**: MIT

Core concept: *Train agent skills like you train neural networks — with epochs, (mini-)batchsize, learning rates, and validation gates — but without touching model weights.*

**Pipeline (6 stages):**
1. **Rollout** — execute episodes with current skill
2. **Reflect** — analyze trajectories, generate patches
3. **Aggregate** — hierarchical merge of patches
4. **Select** — rank and select top edits
5. **Update** — apply edits to skill document
6. **Evaluate** — validate candidate skill, accept/reject

**Output**: `best_skill.md` — a deployable skill document

### Why Integrate with Hermes?

| Hermes Current | With SkillOpt |
|---------------|---------------|
| Skills are manually written SKILL.md files | Skills can be **auto-optimized** from task trajectories |
| Skill improvement requires human iteration | Skill improvement is **data-driven** with validation gates |
| No systematic skill evaluation framework | Built-in **train/val/test split** evaluation |
| Skills are static after creation | Skills can **self-evolve** over time |

#### Option A: CLI Wrapper Skill (Low Effort)
Create a Hermes skill that wraps SkillOpt CLI:
```bash
hermes skillopt train --config configs/hermes/default.yaml --split_dir ./my_tasks
hermes skillopt import ./outputs/best_skill.md --name my-optimized-skill
```

#### Option B: Native Integration (Medium Effort)
- Add SkillOpt as optional dependency
- Provide `hermes skill optimize <skill-name>` command
- Auto-generate training data from Hermes session history
- Integrate with existing skill registry

#### Option C: Deep Integration (High Effort)
- Replace/adapt Hermes skill system to use SkillOpt's document-based skill representation
- Real-time skill optimization during agent execution
- Continuous learning from user feedback

### Technical Compatibility
- ✅ Python 3.10+ (matches Hermes)
- ✅ Supports Azure OpenAI / OpenAI / Anthropic / local vLLM
- ✅ MIT license
- ⚠️ Requires JSON training data in specific format
- ⚠️ Currently benchmark-oriented (SearchQA, ALFWorld, etc.), needs adaptation for general tasks

### Next Steps
1. **Evaluate** SkillOpt on a Hermes skill (e.g., `github-pr-workflow` or `requesting-code-review`)
2. **Prototype** Option A as a proof-of-concept
3. **Gather feedback** from maintainers and community

### References
- [SkillOpt GitHub](https://github.com/microsoft/SkillOpt)
- [SkillOpt Paper](https://arxiv.org/abs/2605.23904)
- [SkillOpt Demo Video](https://youtu.be/JUBMDTCiM0M)

---
*This is a research/experimental feature proposal. Happy to contribute a POC if there's interest!*

### PR DESCRIPTION
**Pull request by fred0m**
## What does this PR do?

Adds **skill-sleep** — a self-contained, validation-gated skill self-improvement pipeline for Hermes. It applies SkillOpt's methodology (rollout → reflect → bounded edit → validation gate → staged adoption) using Hermes's own primitives (CLI, session DB) and `omp` as the optimizer model — **no external dependency on Microsoft's SkillOpt package**.

The pipeline is a standalone tool under `skill-sleep/` with 134 tests, zero external Python dependencies, and full end-to-end verification on real Hermes sessions.

### Architecture

```
skill-sleep.sh (cron entry)
  │
  ├─ [1] MINE      hermes sessions export → friction signals → task cards
  ├─ [2] PROPOSE   omp + muse-spark-1.2 → bounded candidate diff
  ├─ [3] VALIDATE  LLM judge scores each task (gate: threshold 70, pass rate 0.6)
  └─ [4] REVIEW    stage for human review → apply / reject
```

### Key Design Decisions

- **No external package dependency** — pure Python stdlib + `hermes` CLI + `omp`
- **Uses `hermes sessions export --redact`** — stable CLI API, not raw SQLite
- **Bounded edits** — diff capped at 30 added lines (textual learning rate)
- **Rejected buffer** — failed candidates logged to `rejected/` for optimizer context
- **Human-in-the-loop** — REVIEW stage creates a `review.md` with diff + validation scores; apply/reject via CLI

## Related Issue

Closes #32925 (Integrate Microsoft SkillOpt for Self-Evolving Agent Skills)
Relates to #71266 (Native skill-sleep — validation-gated, staged self-improvement)

## Type of Change

- [x] ✨ New feature (non-breaking change that adds functionality)

## Changes Made

- `skill-sleep/README.md` — project documentation
- `skill-sleep/pipeline/mine.py` — MINE: session scanning + friction detection (user corrections, tool errors, retry patterns)
- `skill-sleep/pipeline/propose.py` — PROPOSE: omp optimizer → bounded candidate diff generation
- `skill-sleep/pipeline/validate.py` — VALIDATE: LLM judge scoring + validation gate
- `skill-sleep/pipeline/review.py` — REVIEW: staging, apply, and reject subcommands
- `skill-sleep/lib/task_card.py`, `proposal.py`, `validation.py`, `staging.py` — data models
- `skill-sleep/templates/propose_prompt.md`, `judge_prompt.md` — optimizer and judge prompts
- `skill-sleep/tests/test_mine.py` (23 tests), `test_propose.py` (39 tests), `test_validate.py` (44 tests), `test_review.py` (28 tests)
- `skill-sleep/skill-sleep.sh` — one-shot cron entry point

## How to Test

```bash
# Install omp (if not already installed)
curl -fsSL https://omp.sh/install | sh

# Prerequisites: NINEROUTER_KEY or other omp-compatible provider key
export NINEROUTER_KEY="..."

# Run the full pipeline
cd skill-sleep && bash skill-sleep.sh

# Or run individual stages
python3 -m pipeline.mine --after 7d --output-dir /tmp/test
python3 -m pipeline.propose --tasks /tmp/test/tasks.json --output-dir /tmp/test
python3 -m pipeline.validate --tasks /tmp/test/tasks.json --diff /tmp/test/candidate.diff --proposal /tmp/test/proposal.json --output-dir /tmp/test
python3 -m pipeline.review stage --validation /tmp/test/validation.json --diff /tmp/test/candidate.diff --proposal /tmp/test/proposal.json --output-dir ./

# Run tests
cd skill-sleep && python3 -m pytest tests/ -q
```

### Code

- [x] I've read the [Contributing Guide](https://github.com/NousResearch/hermes-agent/blob/main/CONTRIBUTING.md)
- [x] My commit messages follow [Conventional Commits](https://www.conventionalcommits.org/) (`feat(skill-sleep): ...`)
- [x] I searched for [existing PRs](https://github.com/NousResearch/hermes-agent/pulls) to make sure this isn't a duplicate
- [x] My PR contains **only** changes related to this feature (no unrelated commits)
- [x] I've run `pytest tests/ -q` and all tests pass (134 passed)
- [x] I've added tests for my changes (134 tests across 4 test files)
- [x] I've tested on my platform: macOS 26.3.1 (Apple Silicon)

### Documentation & Housekeeping

- [x] I've updated relevant documentation (README)
- [x] I've considered cross-platform impact (uses only stdlib + subprocess calls; no Windows-specific issues known)
- [ ] N/A: config keys, CONTRIBUTING.md changes, tool descriptions

## Additional Context

This PR was developed in close coordination with the community feature request #71266. The design prioritizes:

1. **Safety** — validation gate prevents skill degradation, staging prevents auto-apply
2. **Simplicity** — zero external deps, pure Python, uses stable Hermes CLI APIs
3. **Transparency** — every stage produces human-readable artifacts (task cards, diffs, validation scores, review.md)

A full end-to-end verification was performed on a real friction episode (remote deployment task with write_file path errors):
- MINE: detected 12 friction signals from 219 sessions
- PROPOSE: generated 9-line candidate diff (remote host pitfalls)
- VALIDATE: scored 92/100 PASS
- REVIEW: staged with complete review.md proposal

### CODE REVIEW

**Comment by alt-glitch** [comment]
> *This was generated by AI during triage.*

Related: #71266, #38753, and #59085 are open self-improving-skill proposals. This separate standalone pipeline overlaps their product scope; a maintainer architecture decision is needed before parallel implementations proceed.
